From fcb2fefb4a27c88d0c9e77c8070cd5cf9d6bbe39 Mon Sep 17 00:00:00 2001
From: Akshay Gupta <Akshay.Gupta@amd.com>
Date: Tue, 7 Apr 2026 13:08:32 +0530
Subject: [PATCH] jtag: jtag-aspeed: Bring changes for Aspeed 26XX

- Bring changes for Aspeed 26XX to the meta driver to support the JTAG
  driver on the Aspeed 2600 chip.
- Aspeed 26xx defines new register offsets which require corresponding
  changes in push_data() and push_data_last().
- The USE_ASPEED_26XX macro gates the new behaviour; disabling it
  restores the previous code path.

Signed-off-by: Akshay Gupta <Akshay.Gupta@amd.com>
---
 drivers/jtag/jtag-aspeed.c | 51 ++++++++++++++++++++++++++++++++++++++
 1 file changed, 51 insertions(+)

diff --git a/drivers/jtag/jtag-aspeed.c b/drivers/jtag/jtag-aspeed.c
index e0a46fd12..a8959096e 100644
--- a/drivers/jtag/jtag-aspeed.c
+++ b/drivers/jtag/jtag-aspeed.c
@@ -20,6 +20,8 @@
 #include <uapi/linux/ioctl.h>
 #include <uapi/linux/jtag.h>
 
+#define USE_ASPEED_26XX
+
 #define ASPEED_SCU_RESET_JTAG		BIT(22)
 #define ASPEED_2600_SCU_CLEAR_REGISTER 0x04
 
@@ -94,6 +96,15 @@
 #define ASPEED_JTAG_GET_TDI(direction, byte) \
 	(((direction) & JTAG_WRITE_XFER) ? byte : UINT_MAX)
 
+/* ASPEED 26XX */
+#define ASPEED_JTAG_CTL_26XX_TRANS_LEN(x)	((x) << 8)
+#define ASPEED_JTAG_TRANS_LEN(len) \
+	(ASPEED_JTAG_CTL_ENG_EN | \
+	 ASPEED_JTAG_CTL_ENG_OUT_EN | \
+	 ASPEED_JTAG_CTL_26XX_TRANS_LEN(len))
+#define ASPEED_JTAG_CTL_26XX_INST_EN		BIT(1)
+#define ASPEED_JTAG_CTL_26XX_LASPEED_TRANS	BIT(4)
+
 #define ASPEED_JTAG_TCK_WAIT		10
 #define ASPEED_JTAG_RESET_CNTR		10
 #define WAIT_ITERATIONS		75
@@ -680,6 +691,23 @@ static int aspeed_jtag_xfer_push_data(struct aspeed_jtag *aspeed_jtag,
 {
 	int res = 0;
 
+#ifdef USE_ASPEED_26XX
+	aspeed_jtag_write(aspeed_jtag, ASPEED_JTAG_TRANS_LEN(bits_len),
+			  ASPEED_JTAG_CTRL);
+	if (type == JTAG_SIR_XFER) {
+		aspeed_jtag_write(aspeed_jtag,
+				  ASPEED_JTAG_TRANS_LEN(bits_len) |
+				  ASPEED_JTAG_CTL_26XX_INST_EN,
+				  ASPEED_JTAG_CTRL);
+		res = aspeed_jtag_wait_instruction_pause(aspeed_jtag);
+	} else {
+		aspeed_jtag_write(aspeed_jtag,
+				  ASPEED_JTAG_TRANS_LEN(bits_len) |
+				  ASPEED_JTAG_CTL_DATA_EN,
+				  ASPEED_JTAG_CTRL);
+		res = aspeed_jtag_wait_data_pause_complete(aspeed_jtag);
+	}
+#else
 	if (type == JTAG_SIR_XFER) {
 		aspeed_jtag_write(aspeed_jtag, ASPEED_JTAG_IOUT_LEN(bits_len),
 				  ASPEED_JTAG_CTRL);
@@ -693,6 +721,7 @@ static int aspeed_jtag_xfer_push_data(struct aspeed_jtag *aspeed_jtag,
 				  ASPEED_JTAG_CTL_DATA_EN, ASPEED_JTAG_CTRL);
 		res = aspeed_jtag_wait_data_pause_complete(aspeed_jtag);
 	}
+#endif
 	return res;
 }
 
@@ -703,6 +732,27 @@ static int aspeed_jtag_xfer_push_data_last(struct aspeed_jtag *aspeed_jtag,
 {
 	int res = 0;
 
+#ifdef USE_ASPEED_26XX
+	aspeed_jtag_write(aspeed_jtag,
+			  ASPEED_JTAG_TRANS_LEN(shift_bits) |
+			  ASPEED_JTAG_CTL_26XX_LASPEED_TRANS,
+			  ASPEED_JTAG_CTRL);
+	if (type == JTAG_SIR_XFER) {
+		aspeed_jtag_write(aspeed_jtag,
+				  ASPEED_JTAG_TRANS_LEN(shift_bits) |
+				  ASPEED_JTAG_CTL_26XX_LASPEED_TRANS |
+				  ASPEED_JTAG_CTL_26XX_INST_EN,
+				  ASPEED_JTAG_CTRL);
+		res = aspeed_jtag_wait_instruction_complete(aspeed_jtag);
+	} else {
+		aspeed_jtag_write(aspeed_jtag,
+				  ASPEED_JTAG_TRANS_LEN(shift_bits) |
+				  ASPEED_JTAG_CTL_26XX_LASPEED_TRANS |
+				  ASPEED_JTAG_CTL_DATA_EN,
+				  ASPEED_JTAG_CTRL);
+		res = aspeed_jtag_wait_data_complete(aspeed_jtag);
+	}
+#else
 	if (type == JTAG_SIR_XFER) {
 		aspeed_jtag_write(aspeed_jtag,
 				  ASPEED_JTAG_IOUT_LEN(shift_bits) |
@@ -726,6 +776,7 @@ static int aspeed_jtag_xfer_push_data_last(struct aspeed_jtag *aspeed_jtag,
 				  ASPEED_JTAG_CTRL);
 		res = aspeed_jtag_wait_data_complete(aspeed_jtag);
 	}
+#endif
 	return res;
 }
 
-- 
2.34.1

